### ISSUE TITLE
AlphaMix+ and others algorithms

### ISSUE THREAD

**Issue by DeepAnonymous** [TradeMaster-NTU/TradeMaster#48]
Is it possible to embed AlphaMix+ into this framework? Or it uses another different framework (MoE).

Also, when I trained EIIE and DeepTrader, the weights are either 0-1 or uniform distributed. I don't know if there is any wrong with the implementation. A quick comparison between the implementation of DeepTrader in this repo and https://github.com/CMACH508/DeepTrader/ shows that the models are different (softmax vs. relu, batch norm, etc.)

I hope you guys can give a quick comments on these issues.

**Comment by DeepAnonymous** [comment]
For DeepTrader, I needed to modify model.py a bit.
`class IN(nn.Module):
    def __init__(self, N, num_features) -> None:
        super(IN, self).__init__()
        self.N = N
        self.linear = nn.Linear(num_features, 1)
        self.bn1 = nn.BatchNorm1d(num_features=num_features)

    def forward(self, S_l, Z_l, H_l):
        x = torch.matmul(S_l, Z_l)
        x = x + H_l
        x = x.reshape(self.N, -1)
        x = self.bn1(x)
        x = self.linear(x)
        x = torch.sigmoid(x).squeeze()
        return x`
I added BN layer (similar to the original paper) to make sure the asset scores are not too large, leading to the biased train weights (only 0 and 1).

Regarding AlphaMix+, I noticed that the weights are mostly diversified with more weights allocated to a few bullish stocks as you said, but do you think it's good (disregard PRUDEX). For example, for a market with >100 stocks, do we really to diversify 100 weights? I think most of them should be 0.

Also, regarding the framework, do you think which one is better TradeMaster or AlphaMix+. Currently, TradeMaster supports both AT and PM, since AlphaMix+ only supports PM.

**Comment by ai-gamer** [comment]
I agree adding BN layer is reasonable. Could you please make a pull request of it?

I think only diversity under good profit is meaningful. Otherwise, we can easily achieve maximum entropy by allocating equal assets into each stock of the portfolio. For a portfolio with large stock pool (e.g., >100), I think an ideal portfolio do not necessarily need to invest money on most of them. Maybe put all money to 10 stock is already enough. However, RL algorithms  such as AlphaMix+ can not take this into account during training at the current stage. We use the heatmap to report the results. If you want a portfolio with more concentration on a few stocks, may simply adding a normalization to set tiny value in the portfolio to 0 is one possible choice.

TradeMaster is developed as a software platform with support of AT, PM and order execution (OE). AlphaMix+ is an RL-based trading algorithm and we conduct experiments on  PM.

**Comment by qinmoelei** [comment]
Thank you for your feedback, we have made some adjustment to our deeptrader model:
for the market score unit, we use exactly the same model as the original code but for the asset scoring unit, since it seems the original code fails to provide correlation matrix among different ticker and therefore they made some  compromise to the model, here we use adjclose correlation among different ticker to do so, and we add the batch norm layer as you just mentioned.

**Comment by DeepAnonymous** [comment]
Nice. I will re-try the code.

Regarding **AlphaMix+** repo (https://github.com/ai-gamer/PRUDEX-Compass/tree/main/AlphaMix%2B/PM), I have one question:

- Since the actions generate +- values (long and short), but we use `softmax_normalization` function to calculate final weights, forcing all the weights > 0. Do you think it's correct? Assuming the whole market going down, but we still allocated positive weights => letting the portfolio losing money. 
- I think it's better to keep +- values or adding 1 for Cash.

For **TradeMaster** repo, I also have one question, not sure whether it's correct.

- In `portfolio_for_deeptrader.py`, should we add **np.abs** when carrying out normalization? 
```
def normalization(self, actions):
        # a normalization function not only for actions to transfer into weights but also for the weights of the
        # portfolios whose prices have been changed through time
        actions = np.array(actions)
        sum = np.sum(**np.abs**(actions))
        actions = actions / sum
        return actions
```

**Comment by ai-gamer** [comment]
I think both of your questions will be solved when the PM environment supporting short(<0) positions. We are working on this and will update it later. Thanks for your suggestions.

**Comment by qinmoelei** [comment]
Dear sir:
Your comment about deeptrader is correct an have updated the related parts.

**Comment by DeepAnonymous** [comment]
Thank for your update. Hopefully, I can make more contribution to the code.

Ultimately, I hope we could embed AlphaMix+ into TradeMaster very soon. Among the PM algorithms, I think AlphaMix+ generate the most reasonable weights. 

Btw, could I confirm that the action space of AlphaMix+ is N+1, and that +1 is for **cash**?
`self.action_space = config["action_space"]+1
...
portfolio_weights = weights[:, 1:]
`
Anyway, the PM environment supporting short(<0) positions is expected.
Thanks

**Comment by ai-gamer** [comment]
Yes, the +1 is for cash.
